File: examples/komegaExamples/kw_o/kw.m
//
//  Kw.m
//  param
//
//  Created by RANDALL JUDD on 2/8/12.
//  Copyright (c) 2012 JVSIP, Independent Consultant. All rights reserved.
//

#import "kw.h"

@implementation Kw

-(id) init
{
   if((self = [super init]))
   {
      Navg = 0;
      Nfreq = 0;
      cm_freq = NULL;
      rm_freq = NULL;
      m_gram = NULL;
      rcfftm = NULL;
      ccfftm = NULL;
      ts_taper = NULL;
      array_taper = NULL;
   }
   return self;
}

-(id) initWithParams:(NSMutableDictionary *)params
{
   vsip_init((void*)0);
   vsip_memory_hint mem=VSIP_MEM_NONE;
   vsip_major col = VSIP_COL;
   vsip_major row = VSIP_ROW;
   vsip_fft_dir fftfwd=VSIP_FFT_FWD;
   vsip_length Nsens=(vsip_length)[[params objectForKey:@"Nsens"] intValue];
   vsip_length Nts = (vsip_length)[[params objectForKey:@"Nts"] intValue];
   if((self = [self init])){
      Navg = (vsip_length) [[params objectForKey:@"Navg"] intValue];
      Nfreq = Nts/2 + 1;
      cm_freq=vsip_cmcreate_f(Nsens,Nfreq,col,mem);
      rm_freq=vsip_mcreate_f(Nsens,Nts,row,mem);
      m_gram=vsip_mcreate_f(Nsens,Nfreq,col,mem);
      rcfftm=vsip_rcfftmop_create_f(Nsens,Nts,1,row,0,0);
      ccfftm=vsip_ccfftmip_create_f(Nsens,Nfreq,fftfwd,1,col,0,0);
      ts_taper=vsip_vcreate_hanning_f(Nts, mem);
      array_taper=vsip_vcreate_hanning_f(Nsens,0);
   }
   return self;
}

-(void) dealloc
{
   if(cm_freq) vsip_cmalldestroy_f(cm_freq);
   if(rm_freq) vsip_malldestroy_f(rm_freq);
   if(m_gram) vsip_malldestroy_f(m_gram);
   if(ts_taper) vsip_valldestroy_f(ts_taper);
   if(array_taper) vsip_valldestroy_f(array_taper);
   if(rcfftm) vsip_fftm_destroy_f(rcfftm);
   if(ccfftm) vsip_fftm_destroy_f(ccfftm);
   vsip_finalize((void*)0);
   [super dealloc];
}

-(void) zero
{
   vsip_mfill_f(0.0, m_gram);
}

-(void) kw:(vsip_mview_f*)m_data
{
   vsip_major row = VSIP_ROW;
   vsip_major col = VSIP_COL;
   vsip_vmmul_f(ts_taper, m_data, row, m_data);
   vsip_vmmul_f(array_taper,m_data,col,m_data);
   vsip_rcfftmop_f(rcfftm, m_data, cm_freq);
   vsip_ccfftmip_f(ccfftm, cm_freq);
   vsip_mcmagsq_f(cm_freq, rm_freq);
   vsip_smmul_f(1.0/(vsip_scalar_f)Navg,rm_freq,rm_freq);
   vsip_madd_f(rm_freq,m_gram,m_gram);
}

-(vsip_mview_f*) instance
{
   return m_gram;
}


@end

File: examples/komegaExamples/kw_py/kw.py
import vsiputils as vsip
class Kw(object):
    def __init__(self,param):
        mem=vsip.VSIP_MEM_NONE
        col=vsip.VSIP_COL
        row=vsip.VSIP_ROW
        fftfwd=vsip.VSIP_FFT_FWD
        vsip.init()
        Nsens=param['Nsens']
        Nts=param['Nts']
        Nfreq=int(Nts/2) + 1
        self.Navg=param['Navg']
        self.Nfreq=Nfreq
        self.cm_freq=vsip.create('cmview_d',(Nsens,Nfreq,col,mem)) 
        self.rm_freq=vsip.realview(self.cm_freq)
        self.m_gram=vsip.create('mview_d',(Nsens,Nfreq,col,mem))
        self.rcfftm=vsip.create('rcfftmop_d',(Nsens,Nts,1,row,0,0))
        self.ccfftm=vsip.create('ccfftmip_d',(Nsens,Nfreq,fftfwd,1,col,0,0))
        self.ts_taper=vsip.create('hanning_d',(Nts,0))
        self.array_taper=vsip.create('hanning_d',(Nsens,0));

    def __del__(self):
        vsip.destroy(self.rm_freq)
        vsip.allDestroy(self.cm_freq)
        vsip.allDestroy(self.m_gram)
        vsip.allDestroy(self.ts_taper)
        vsip.allDestroy(self.array_taper)
        vsip.destroy(self.rcfftm)
        vsip.destroy(self.ccfftm)
        vsip.finalize()

    def zero(self):
        vsip.fill(0.0,self.m_gram)

    def kw(self,m_data):
        row=vsip.VSIP_ROW
        col=vsip.VSIP_COL
        vsip.vmmul(self.ts_taper,m_data,row,m_data);
        vsip.vmmul(self.array_taper,m_data,col,m_data);
        vsip.fftop(self.rcfftm,m_data,self.cm_freq)
        vsip.fftip(self.ccfftm,self.cm_freq)
        vsip.magsq(self.cm_freq,self.rm_freq) #check rm_freq real part?
        vsip.mul(1.0/self.Navg,self.rm_freq, self.rm_freq)
        vsip.add(self.rm_freq,self.m_gram,self.m_gram)

    def instance(self):
        return self.m_gram
